server/mdm/apple/cert.go

```diff
@@ -0,0 +1,123 @@
+package apple_mdm
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/micromdm/scep/v2/depot"
+)
+
+const defaultFleetDMAPIURL = "https://fleetdm.com"
+
+const getSignedAPNSCSRPath = "/api/v1/get_signed_apns_csr"
+
+// emailAddressOID defined by https://oidref.com/1.2.840.113549.1.9.1
+var emailAddressOID = []int{1, 2, 840, 113549, 1, 9, 1}
+
+// GenerateAPNSCSRKey generates a APNS csr to be sent to fleetdm.com and returns a csr and key.
+func GenerateAPNSCSRKey(email, org string) (*x509.CertificateRequest, *rsa.PrivateKey, error) {
+	key, err := NewPrivateKey()
+	if err != nil {
+		return nil, nil, fmt.Errorf("generate private key: %w", err)
+	}
+
+	subj := pkix.Name{
+		Organization: []string{org},
+		ExtraNames: []pkix.AttributeTypeAndValue{{
+			Type:  emailAddressOID,
+			Value: email,
+		}},
+	}
+	template := &x509.CertificateRequest{
+		Subject:            subj,
+		SignatureAlgorithm: x509.SHA256WithRSA,
+	}
+
+	b, err := x509.CreateCertificateRequest(rand.Reader, template, key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certReq, err := x509.ParseCertificateRequest(b)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return certReq, key, nil
+}
+
+type getSignedAPNSCSRRequest struct {
+	// CSR is the pem encoded certificate request.
+	CSR []byte `json:"csr"`
+}
+
+func GetSignedAPNSCSR(client *http.Client, csr *x509.CertificateRequest) error {
```

Comment:
Nit, but that's a big confusing reading this as it's called `GetSignedAPNSCSR` but it does not return anything (but an error). I think it's because the request it makes will send an email with the signed CSR on success, but it would benefit from having a comment explaining that.

Reply:
Yeah, I agree it's a bit confusing. Added a comment in 5a2af17